src/lsm/table_memory.zig

```diff
@@ -133,176 +362,187 @@ pub fn TableMemoryType(comptime Table: type) type {
+                dedup_sink.push(value);
+            }
+            table_immutable.value_context.count = dedup_sink.finish();
+
+            table_immutable.finalize(snapshot_min);
+
+            assert(table_immutable.sorted());
+
+            table_mutable.reset();
+        }
+
+        // Absorb the current immutable table into the mutable one,
+        // then re-materialize a compact immutable table.
+        pub fn absorb(
+            table_immutable: *TableMemory,
+            table_mutable: *TableMemory,
+            snapshot_min: u64,
+        ) void {
+            defer assert(table_immutable.sorted());
+            assert(table_immutable.mutability == .immutable);
+            maybe(table_immutable.mutability.immutable.absorbed);
+            assert(table_immutable.sorted());
+            assert(table_mutable.mutability == .mutable);
+            maybe(table_mutable.sorted());
+
             const values_count_limit = table_immutable.values.len;
-            assert(values_count_limit == values_count_limit);
             assert(table_immutable.count() <= values_count_limit);
             assert(table_mutable.count() <= values_count_limit);
             assert(table_immutable.count() + table_mutable.count() <= values_count_limit);
 
+            if (table_mutable.count() == 0) {
+                table_mutable.reset();
+                return;
+            }
+
+            // Copy immutable after the current mutable tail and mark as an immutable run,
+            // so the merge prefers its entries on key ties.
+            const tables_combined_count = table_immutable.count() + table_mutable.count();
             stdx.copy_disjoint(
                 .inexact,
                 Value,
-                table_immutable.values[table_immutable.count()..],
-                table_mutable.values[0..table_mutable.count()],
+                table_mutable.values[table_mutable.count()..],
+                table_immutable.values[0..table_immutable.count()],
             );
-
-            const tables_combined_count = table_immutable.count() + table_mutable.count();
-            table_immutable.value_context.count =
-                sort_suffix_from_offset(table_immutable.values[0..tables_combined_count], 0);
-            assert(table_immutable.count() <= tables_combined_count);
-
-            table_mutable.reset();
-            table_immutable.mutability = .{ .immutable = .{
-                .flushed = table_immutable.value_context.count == 0,
-                .absorbed = true,
-                .snapshot_min = snapshot_min,
-            } };
+            table_mutable.value_context.run_tracker.add(.{
+                .min = table_mutable.count(),
+                .max = tables_combined_count,
+                .origin = .immutable,
+            });
+            table_mutable.value_context.count = tables_combined_count;
+
+            table_immutable.compact(table_mutable, snapshot_min);
```

Comment:
After compact can we assert:

```suggestion
            table_immutable.compact(table_mutable, snapshot_min);

            assert(table_immutable.value_context.run_tracker.count() == 1);
            assert(table_mutable.value_context.run_tracker.count() == 0);
```